src/sources/udp.rs

```diff
@@ -0,0 +1,220 @@
+use crate::{
+    event::{self, Event},
+    topology::config::{DataType, GlobalOptions, SourceConfig},
+};
+use codec::{self, BytesDelimitedCodec};
+use futures::{future, stream::iter_ok, sync::mpsc, Future, Sink, Stream};
+use serde::{Deserialize, Serialize};
+use std::{io, net::SocketAddr};
+use string_cache::DefaultAtom as Atom;
+use tokio::{
+    self,
+    codec::{BytesCodec, Decoder},
+    net::{UdpFramed, UdpSocket},
+};
+use tracing::field;
+
+/// UDP processes messages per packet, where messages are separated by newline.
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct UdpConfig {
+    pub address: SocketAddr,
+    pub host_key: Option<Atom>,
+}
+
+impl UdpConfig {
+    pub fn new(address: SocketAddr) -> Self {
+        Self {
+            address,
+            host_key: None,
+        }
+    }
+}
+
+#[typetag::serde(name = "udp")]
+impl SourceConfig for UdpConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> Result<super::Source, String> {
+        let host_key = self.host_key.clone().unwrap_or(event::HOST.clone());
+        Ok(udp(self.address, host_key, out))
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+pub fn udp(address: SocketAddr, host_key: Atom, out: mpsc::Sender<Event>) -> super::Source {
+    let out = out.sink_map_err(|e| error!("error sending event: {:?}", e));
+
+    Box::new(
+        future::lazy(move || {
+            let socket = UdpSocket::bind(&address).expect("failed to bind to udp listener socket");
+
+            info!(message = "listening.", addr = &field::display(address),);
+
+            future::ok(socket)
+        })
+        .and_then(move |socket| {
+            let host_key = host_key.clone();
+            let mut new_line_decoder = BytesDelimitedCodec::new(b'\n');
+            UdpFramed::new(socket, BytesCodec::new())
```

Comment:
I believe BytesDelimitedCodec should still work here since it won't drain the entire BytesMut.

Reply:
Yep, and to be clear that means we should be able to skips the parts below where we manually pull messages through and stick them in a `Vec`. The `UdpFramed` should let us go straight to a stream of messages.

Reply:
Indeed `BytesDelimitedCodec` works, but in junction with `UdpFramed`. `UdpFramed` has `poll` method which receives a whole datagram, and then calls decoder with it. The key thing is that it calls decoder only once before clearing the buffer that contains the datagram, and with it all but the first message is lost. Also `poll` calls `decode`, but `decode_eof` is needed.

Although I could implement a more appropriate version of `UdpFramed` to avoid that while loop.

Reply:
A more simpler solution has been committed. 
I just turned the while loop into a stream. 
That eliminated the use of `Vec`.

Reply:
Ah, good point. Looks like we handle this a little differently between the syslog udp source and the statsd udp source. I think what you have here makes sense.

Reply:
Great, I'll close this then.